index.js

```diff
@@ -1,18 +1,63 @@
 'use strict';
 
-module.exports = function implicitFiguresPlugin(md) {
+var figure = function (options) {
 
-  var defaultRender = md.renderer.rules.image;
+  var dataType = (typeof options.dataType !== 'undefined') ? options.dataType : false;
 
-  md.renderer.rules.image = function (tokens, idx, options, env, self) {
-    var token = tokens[idx];
-    var img = defaultRender(tokens, idx, options, env, self);
+  return function implicit_figure (state, startLine, endLine, silent) {
+    var content, terminate, i, l, token,
+        prevLine = startLine - 1,
+        nextLine = startLine + 1,
+        terminatorRules = state.md.block.ruler.getRules('implicit_figure'),
+        endLine = state.lineMax,
+        openTag = dataType ? 'figure data-type="image"' : 'figure';
 
-    if (tokens.length === 1) {
-      // image is alone
-      img = '<figure>'+ img +'</figure>';
+    // jump line-by-line until empty one or EOF
+    for (; nextLine < endLine && !state.isEmpty(nextLine); nextLine++) {
+      // this would be a code block normally, but after paragraph
+      // it's considered a lazy continuation regardless of what's there
+      if (state.sCount[nextLine] - state.blkIndent > 3) { continue; }
+
+      // quirk for blockquotes, this line should already be checked by that rule
+      if (state.sCount[nextLine] < 0) { continue; }
+
+      // Some tags can terminate paragraph without empty line.
+      terminate = false;
+      for (i = 0, l = terminatorRules.length; i < l; i++) {
+        if (terminatorRules[i](state, nextLine, endLine, true)) {
+          terminate = true;
+          break;
+        }
+      }
+      if (terminate) { break; }
     }
 
-    return img;
-  }
+    if (!state.isEmpty(prevLine) || !state.isEmpty(nextLine)) { return false; }
+
+    content = state.getLines(startLine, nextLine, state.blkIndent, false).trim();
+
+    if (content.charCodeAt(0) !== 0x21/* ! */) { return false; }
+    if (content.charCodeAt(1) !== 0x5B/* [ */) { return false; }
```

Comment:
Instead of reimplementing the parser, another approach is [replacing the tokens](https://github.com/arve0/markdown-it-header-sections/blob/master/index.js).